localstack/dns/server.py

```diff
@@ -0,0 +1,875 @@
+                if record_type == RecordType.AAAA:
+                    ip = "::1"
+                return Record(RDMAP.get(record_type.name), ip)
+            case _:
+                raise NotImplementedError(f"Record type '{type(name_record)}' not implemented")
+
+    def _determine_best_ip(self) -> str:
+        client_ip, _ = self.client_address
+        # allow for overriding if required
+        if config.DNS_RESOLVE_IP != LOCALHOST_IP:
+            return config.DNS_RESOLVE_IP
+
+        # Look up best matching ip address for the client
+        interfaces = self._fetch_interfaces()
+        for interface in interfaces:
+            subnet = interface.network
+            ip_address = IPv4Address(client_ip)
+            if ip_address in subnet:
+                # check if the request has come from the gateway or not. If so
+                # assume the request has come from the host, and return
+                # 127.0.0.1
+                if config.is_in_docker and self._is_gateway(ip_address):
+                    return LOCALHOST_IP
+
+                return str(interface.ip)
+
+        # no best solution found
+        LOG.warning(
+            f"could not determine subnet-matched IP address for {self.request.q.qname}, falling back to {LOCALHOST_IP}"
+        )
+        return LOCALHOST_IP
+
+    @staticmethod
+    def _is_gateway(ip: IPv4Address) -> bool:
+        """
+        Look up the gateways that this contianer has, and return True if the
+        supplied ip address is in that list.
+        """
+        return ip == iputils.get_default_gateway()
+
+    @staticmethod
+    def _fetch_interfaces() -> Iterable[IPv4Interface]:
+        interfaces = list_network_interface_details()
+        for _, addresses in interfaces.items():
+            for address in addresses:
+                if address.family != AddressFamily.AF_INET:
+                    # TODO: IPv6
+                    continue
+
+                # argument is of the form e.g. 127.0.0.1/255.0.0.0
+                net = IPv4Interface(f"{address.address}/{address.netmask}")
+                yield net
+
+
+class NonLoggingHandler(DNSHandler):
+    """Subclass of DNSHandler that avoids logging to stdout on error"""
+
+    def handle(self, *args, **kwargs):
+        try:
+            THREAD_LOCAL.client_address = self.client_address
+            THREAD_LOCAL.server = self.server
+            THREAD_LOCAL.request = self.request
+            return super(NonLoggingHandler, self).handle(*args, **kwargs)
+        except Exception:
+            pass
+
+
+SKIP_PATTERNS = {
+    r".*(forums|console|docs|clientvpn|sso|boto3|(signin(\-reg)?))\.([^\.]+\.)?(aws\.amazon|amazonaws)\.com",
+    r".*captcha-prod\.s3\.amazonaws\.com",
+    r"^aws\.amazon\.com",
+    r"^github-production-release-.*\.s3\.amazonaws\.com",
+    r"^aws-glue-etl-artifacts\.s3\.amazonaws\.com",
+    rf"^{S3_ASSETS_BUCKET}\.s3\.amazonaws\.com",
+    r"^localstack-pods-.*\.s3\.amazonaws\.com",
+}
```

Comment:
I would argue if this should be here. Since we will add the transparent endpoint injection in -ext, this might fit better there as well. By default, we would not redirect these entries anyway.